src/task_manager/preemptive.rs

```diff
@@ -0,0 +1,127 @@
+use crate::ports::{Port, PortTrait, TrapFrame, STACK_ALIGN};
+use crate::task_manager::task::{
+    Task, TaskLoopFunctionType, TaskSetupFunctionType, TaskStopConditionFunctionType,
+};
+use crate::task_manager::{TaskManagerTrait, TASK_MANAGER};
+use alloc::vec::Vec;
+use core::alloc::Layout;
+use core::sync::atomic::{AtomicUsize, Ordering};
+
+pub(crate) const THREAD_STACK_SIZE: usize = 1024; // TODO:
+
+static mut NEXT_THREAD_ID: AtomicUsize = AtomicUsize::new(0);
+
+pub(crate) struct Thread {
+    /// id of this thread
+    id: usize,
+    /// Pointer to the memory allocated for stack
+    pub(crate) stack: *mut u8,
+    /// **Arch specific** state of the registers at the moment of context switch
+    pub(crate) context: TrapFrame,
+    /// Task that is executed by this thread
+    pub(crate) task: Task,
+}
```

Comment:
Maybe it would be more logical to extend task rather than add a thread to it? As if it can be confusing when there are these two entities.

Reply:
Isn't it more or less the same situation as in the case of `Task` and `FutureTask` structures, where the former represents what the task should do and the latter represents an additional context for the cooperative scheduler?

Reply:
Ok, fair enough